### ISSUE TITLE
Typechecking on mutually recursive types is buggy when used in a union with Symbol

### ISSUE THREAD

**Issue by lexi-lambda** [racket/typed-racket#26]
> Note: This bug is so strange that it's rather difficult to even reproduce. I'll try and describe exactly what circumstances allow me to reproduce it.

Take the following trivial program. Call it `bug.rkt`.

```racket
#lang typed/racket/base

(define-type Bson-Document (HashTable Symbol Bson-Value))
(define-type Bson-Value (U Bson-Document Symbol))

(define (make-document) : Bson-Value
  (define log-data : Bson-Document (make-immutable-hasheq `()))
  log-data)
```

For me, if I run `raco setup`, then running `racket bug.rkt` typechecks just fine, as it should! However, now I run the following command:

    raco setup -c --pkgs typed-racket-more

This shouldn't have any affect on any program, much less this one, which uses absolutely *zero* functionality from `typed-racket-more`. Nevertheless, now if I run `racket bug.rkt`, I get the following type error:

```
/tmp/bug.rkt:8:2: Type Checker: type mismatch
  expected: Bson-Value
  given: Bson-Document
  in: log-data
  context...:
   /Applications/Racket v6.1.1.8/share/pkgs/typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt:284:0: type-check
   /Applications/Racket v6.1.1.8/share/pkgs/typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt:494:0: tc-module
   /Applications/Racket v6.1.1.8/share/pkgs/typed-racket-lib/typed-racket/tc-setup.rkt:39:0: tc-setup
   /Applications/Racket v6.1.1.8/share/pkgs/typed-racket-lib/typed-racket/typed-racket.rkt:18:4
   standard-module-name-resolver
```

Don't worry, it gets weirder. I now tried modifying the definition for `Bson-Value`. I changed it to `(U Bson-Document String)`—anything that didn't include `Symbol` worked. Then I ran `racket bug.rkt` again. No error!

Okay... now I changed `Bson-Value` back to `(U Bson-Document Symbol)`. This time, though, I reversed the order of the type type declarations, so lines 3 and 4 looked like this:

```racket
(define-type Bson-Value (U Bson-Document Symbol))
(define-type Bson-Document (HashTable Symbol Bson-Value))
```

Again, no error! **Wat.** Now utterly confused, I ran `raco setup --pkgs typed-racket-lib`. Then I ran `racket bug.rkt`, unchanged since I swapped the lines. Guess what? Type error!

I then went back to the original code I posted at the top of this issue and ran it, but this time, it typechecked fine. I'm so hopelessly confused.

---

What the hell is going on? I don't know. Here's how I've managed to reproduce this bug.

1. I've attempted this on a nice, clean Ubuntu 14.04 VM. I've installed [64-bit Racket from the utah.edu snapshots page](http://www.cs.utah.edu/plt/snapshots/).
2. I've run `raco setup` followed by `raco setup -c --pkgs typed-racket-more`.
3. I've run the originally posted `bug.rkt`.

Those steps give me an error. As for all the other bizarre oddities... I have no idea. Try it for yourself. Let me know what you find. I'm baffled.

**Comment by samth** [comment]
@lexi-lambda tried this with HEAD of `typed-racket` and v6.1.1 of Racket, and it didn't happen, which suggests that it's a Racket bug.

**Comment by samth** [comment]
Both @mflatt and I have been having trouble reproducing this. @takikawa, you reproduced this, right? Do you have a recipe I could follow?

**Comment by takikawa** [comment]
Yes. The `raco setup -c --pkgs typed-racket-more` step does not seem necessary. BTW, I think this command deletes too much.

Instead, just delete `typed/racket/base/compiled` and then run `racket bug.rkt`. For me, it does not error in `--no-jit` mode.

**Comment by takikawa** [comment]
BTW, just to update anyone else who was keeping track of this, I think that the bug is in TR and in particular the mutually recursive type alias part. The following program illustrates the bug without relying on particular hash orders:

````
#lang typed/racket

(define-type T1 (Listof (U T2 String)))
(define-type T2 (Vectorof (U T1 String)))

(: x1 T1)
(define x1 (list (vector "foo")))

(: x2 T2)
(define x2 (vector (list "foo")))
````

I think this will fail to typecheck no matter what hash ordering is picked, though I haven't been able to confirm this yet because I can only get a type error referring to `x1` and `T1` so far. It should typecheck though.

A potential solution is to change the algorithm in TR to compute a fixed-point instead of only doing two passes to register these type aliases. Though we may also need to disable subtype caching temporarily during the fixed-point too (not sure yet).

**Comment by lexi-lambda** [comment]
The problem mentioned by @takikawa seems like a real one, but I think it might not ultimately be the source of this particular problem. I spent a few hours tracing the path the typechecking took through the code, and I found this.

See [this line of type-alias-helper.rkt](https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/env/type-alias-helper.rkt#L94). Basically, it registers all the type aliases with a temporary placeholder value to be substituted with the real information later. It does this by using `(make-Value (gensym))` as a dummy value.

This *seems* like an okay idea, but it's actually problematic. In this case, the typechecker then attempts to parse the type `(U Bson-Document Symbol)`. The `Un` constructor is called, which eliminates subtypes so that, for example, `(U Number Integer)` would be resolved to simply `Number`.

Now the problem becomes apparent. At this point in time, it resolves `Bson-Document` to its type, which is currently `(-val (gensym))`, and obviously, a symbol *is in fact a subtype* of `Symbol`!

Therefore, the type of `Bson-Value` is collapsed to simply `Symbol`, and the typechecker happily carries on. Clearly, this is a bug.

As for a solution, as suggested by @takikawa on IRC:

> I think what's needed is basically a fixpoint there instead of two fixed passes. Keep re-parsing until the types stop changing.

This would fix this problem, and it would also fix the more general issue mentioned above. If it's an easy fix, great! I'd appreciate seeing it implemented. If not, I'd also appreciate a workaround for this particular issue. The trouble is really the use of placeholder types that are symbol values. Is there some type that fails all subtyping checks?

**Comment by takikawa** [comment]
I think you're right @lexi-lambda that that code is suspicious, but it also turns out that my example was wrong. The example I gave above shouldn't typecheck because of the invariance of the `Vectorof` constructor. A better example is this one:

````
#lang typed/racket

(define-type T1 (Listof (U T2 String)))
(define-type T2 (Setof (U T1 String)))

(: x1 T1)
(define x1 (list (set "foo")))

(: x2 T2)
(define x2 (set (list "foo")))
````

This snippet, though, does not seem to have any trouble typechecking. So I think my earlier assessment was wrong.

**Comment by takikawa** [comment]
Sorry, it looks like I had a brainfart here. The right example is actually:

````
#lang typed/racket

(define-type T1 (Listof (U T2 Symbol)))
(define-type T2 (Setof (U T1 Symbol)))

(: x1 T1)
(define x1 (list (set 'foo)))

(: x2 T2)
(define x2 (set (list 'foo)))
````

which does fail in the same way that @lexi-lambda points out.

**Comment by lexi-lambda** [comment]
In that case, it would seem that the solution would be to replace `(-val (gensym))` with a placeholder type that is *never* the supertype or subtype of any other type. On IRC, @takikawa mentioned surrounding it with `(-box ...)`, but then realized that could still collapse with `BoxTop`.

Is there any such type?

**Comment by takikawa** [comment]
A candidate is a `(make-Base 'Placeholder ...)` since I think base types are not subtypes of anything else (except `Any`) unless there's an explicit special case in subtyping. They also should only be a supertype of `Bottom`.

**Comment by takikawa** [comment]
The immediate issue here was closed by the fix in 619d694. I just wanted to note that PR 14549 (http://bugs.racket-lang.org/query/?debug=&database=default&cmd=view+audit-trail&cmd=view&pr=14549) is a related bug that isn't completely fixed by this change.

In particular, the example in PR 14549 will succeed/fail based on seemingly random differences now (like whether the test is run standalone or in the TR test suite).

Just for the record, here's a proposed solution for the issue by @mithos28

> The solution that I propose is that trying to resolve a unresolved
> type alias is an error. Then we need to fix all the call sites of
> resolve to additionally handle this case (if they could possibly be
> called). Subtype and infer would be the big offenders and the
> additional code is easy there (just return #f). We would need a way to
> make sure that subtyping didn't write into the cache at that point
> though. I'm not sure about other cases or total fallout of this
> change.

I think that disabling the subtype cache during the initial type alias parsing may also work.
